lightning/src/ln/onion_utils.rs

```diff
@@ -1474,6 +1468,250 @@ where
+	InvalidTrampolinePayload,
+	/// A payment was rejected because it did not include the correct payment secret from an
+	/// invoice.
+	PaymentSecretRequired,
+	/// The HTLC was failed because its expiry is too close to the current block height, and we
+	/// expect that it will immediately be failed back by our downstream peer.
+	OutgoingCLTVTooSoon,
+	/// The HTLC was failed because it was pending on a channel which is now in the process of
+	/// being closed.
+	ChannelClosed,
+	/// The HTLC was failed back because its expiry height was reached and funds were timed out
+	/// on chain.
+	OnChainTimeout,
+	/// UnknownFailureCode represents BOLT04 failure codes that we are not familiar with. We will
+	/// encounter this if:
+	/// - A peer sends us a new failure code that LDK has not yet been upgraded to understand.
+	/// - We read a deprecated failure code from disk that LDK no longer uses.
+	///
+	/// See <https://github.com/lightning/bolts/blob/master/04-onion-routing.md#returning-errors>
+	/// for latest defined error codes.
+	UnknownFailureCode {
+		/// The bolt 04 failure code.
+		code: u16,
+	},
+}
+
+impl LocalHTLCFailureReason {
+	pub(super) fn failure_code(&self) -> u16 {
+		match self {
+			Self::TemporaryNodeFailure | Self::ForwardExpiryBuffer => NODE | 2,
+			Self::PermanentNodeFailure => PERM | NODE | 2,
+			Self::RequiredNodeFeature | Self::PaymentSecretRequired => PERM | NODE | 3,
+			Self::InvalidOnionVersion => BADONION | PERM | 4,
+			Self::InvalidOnionHMAC => BADONION | PERM | 5,
+			Self::InvalidOnionKey => BADONION | PERM | 6,
+			Self::TemporaryChannelFailure
+			| Self::DustLimitHolder
+			| Self::DustLimitCounterparty
+			| Self::FeeSpikeBuffer
+			| Self::ChannelNotReady => UPDATE | 7,
+			Self::PermanentChannelFailure | Self::OnChainTimeout | Self::ChannelClosed => PERM | 8,
+			Self::RequiredChannelFeature => PERM | 9,
+			Self::UnknownNextPeer
+			| Self::PrivateChannelForward
+			| Self::RealSCIDForward
+			| Self::InvalidTrampolineForward => PERM | 10,
+			Self::AmountBelowMinimum => UPDATE | 11,
+			Self::FeeInsufficient => UPDATE | 12,
+			Self::IncorrectCLTVExpiry => UPDATE | 13,
+			Self::CLTVExpiryTooSoon | Self::OutgoingCLTVTooSoon => UPDATE | 14,
+			Self::IncorrectPaymentDetails
+			| Self::PaymentClaimBuffer
+			| Self::InvalidKeysendPreimage => PERM | 15,
+			Self::FinalIncorrectCLTVExpiry => 18,
+			Self::FinalIncorrectHTLCAmount => 19,
+			Self::ChannelDisabled => UPDATE | 20,
+			Self::CLTVExpiryTooFar => 21,
+			Self::InvalidOnionPayload | Self::InvalidTrampolinePayload => PERM | 22,
+			Self::MPPTimeout => 23,
+			Self::InvalidOnionBlinding => BADONION | PERM | 24,
+			Self::UnknownFailureCode { code } => *code,
+		}
+	}
+
+	pub(super) fn is_temporary(&self) -> bool {
+		self.failure_code() & UPDATE == UPDATE
+	}
+
+	#[cfg(test)]
+	pub(super) fn is_permanent(&self) -> bool {
+		self.failure_code() & PERM == PERM
+	}
+}
+
+impl Into<LocalHTLCFailureReason> for u16 {
```

Comment:
Prefer `From` over `Into` as per the `Into` docs: https://doc.rust-lang.org/std/convert/trait.Into.html